src/renderer/components/file/file.tsx

```diff
@@ -0,0 +1,188 @@
+    maxSize?: number;
+    // default is "Reject"
+    onOverSizeLimit?: OverSizeLimitStyle;
+    
+    maxTotalSize?: number;
+    // default is "Reject"
+    onOverTotalSizeLimit?: OverTotalSizeLimitStyle;
+}
+
+export interface State {
+    errorText?: string;
+    status: "CLEAR" | "PROCESSING" | "ERROR";
+}
+
+export type Props = BaseProps & (MemoryUseProps | FileUploadProps) ;
+
+export class FilePicker extends React.Component<Props, State> {
+  static defaultProps = {
+    maxSize: Infinity,
+    onOverSizeLimit: "Reject",
+    maxTotalSize: Infinity,
+    onOverLimit: "Reject",
+    onOverTotalSizeLimit: "Reject",
+  }
+
+  constructor(props: Props) {
+    super(props);
+    this.state = {
+      status: "CLEAR",
+    };
+  }
+
+  handleFileCount(files: File[]): File[] {
+    const { limit: [minLimit, maxLimit] = [0, Infinity], onOverLimit } = this.props;
+    if (files.length > maxLimit) {
+      switch (onOverLimit) {
+      case "Cap":
+        files.length = maxLimit;
+        break;
+      case "Reject":
+        throw `Too many files. Expected at most ${maxLimit}. Got ${files.length}.`;
+      }
+    }
+    if (files.length < minLimit) {
+      throw `Too many files. Expected at most ${maxLimit}. Got ${files.length}.`;
+    }
+
+    return files;
+  }
+
+  handleIndiviualFileSizes(files: File[]): File[] {
+    const { onOverSizeLimit, maxSize } = this.props;
+
+    switch (onOverSizeLimit) {
+    case "Filter":
+      return files.filter(file => file.size <= maxSize );
+    case "Reject":
+      const firstFileToLarge = files.find(file => file.size > maxSize);
+      if (firstFileToLarge) {
+        throw `${firstFileToLarge.name} is too large. Maximum size is ${maxSize}. Has size of ${firstFileToLarge.size}`;
+      }
+
+      return files;
+    }
+  }
+
+  handleTotalFileSizes(files: File[]): File[] {
+    const { maxTotalSize, onOverTotalSizeLimit } = this.props;
+
+    const totalSize = files.map(f => f.size).reduce((a, b) => a + b, 0);
+    if (totalSize <= maxTotalSize) {
+      return files;
+    }
+
+    switch (onOverTotalSizeLimit) {
+    case "FilterLargest":
+      files = files.sort((a, b) => a.size - b.size);
+    case "FilterLast":
+      let newTotalSize = totalSize;
+      
+      for (;files.length > 0;) {
+        newTotalSize -= files.pop().size;
+        if (newTotalSize <= maxTotalSize) {
+          break;
+        }
+      }
+      return files;
+    case "Reject":
+      throw `Total file size to upload is too large. Expected at most ${maxTotalSize}. Found ${totalSize}.`;
+    }
+  }
+
+  async handlePickFiles(selectedFiles: FileList) {
+    const files: File[] = Array.from(selectedFiles);
+
+    try {
+      const numberLimitedFiles = this.handleFileCount(files);
+      const sizeLimitedFiles = this.handleIndiviualFileSizes(numberLimitedFiles);
+      const totalSizeLimitedFiles = this.handleTotalFileSizes(sizeLimitedFiles);
+      
+      if ("uploadDir" in this.props) {
```

Comment:
Why not `if(this.props.uploadDir)`?

Reply:
But you get the error `"uploadDir" does not exist on (Readonly<BaseProps & FileUploadProps> & Readonly<{ children?: ReactNode; }>) | (Readonly<BaseProps & MemoryUseProps> & Readonly<...>)`

Reply:
Well, question is why it's not exists there? you define shape of the props for this component?

Reply:
Why is it discouraged?

Reply:
> The `in` operator returns true if the specified property is in the specified object or its prototype chain.
because you might know/define exact shape of `props` and you don't need to "guess"

Reply:
I guess then I should use `this.props.hasOwnProperty("uploadDir")`. The reason that it is not always defined is deliberate. The type of `handler` is dependent on whether `uploadDir` is present. Namely see `FileUploadProps` vs `MemoryUseProps`.

Reply:
So why not define as `optional` prop and check normally?